### ISSUE TITLE
More variance in metrics on Linux version than old Windows version?

### ISSUE THREAD

**Issue by soulgalore** [WPO-Foundation/wptagent#79]
Hi Pat,
I wanted to check if it is only we or if you also seen this:

We run 5 runs per URL, we pick the fastest of the runs with Speed Index (following how you at least used to do with [testing Chrome](https://calendar.perfplanet.com/2015/working-on-chrome-performance-with-webpagetest/)). When we first switched to fastest (from median) I clearly remember it gave us more stable metrics (but I cannot find my notes about it).

Here's what it looks like https://en.wikipedia.org/wiki/Sweden for Chrome on Windows:
<img width="1263" alt="screen shot 2018-02-07 at 9 28 03 am" src="https://user-images.githubusercontent.com/540757/35905981-42268164-0be9-11e8-8221-bd41c83eb3de.png">

On Windows the diff is 200 points for SpeedIndex (12.5%). 

And here on the Linux instance:
<img width="1258" alt="screen shot 2018-02-07 at 9 29 04 am" src="https://user-images.githubusercontent.com/540757/35906020-6a60cc3e-0be9-11e8-972f-49a82ce8d6ff.png">

For Linux the diff is 464 points for SpeedIndex (56%).  Yep on Linux we render faster but the diff in points is much larger too.

We have the same pattern on all the URLs we test. We run both on C4.large. We run default setup so I think for Linux then we have 30 fps for the video. Do you think we should increase the instance size to get less variable metrics?

I'm gonna change to pick the median value (instead of fastest) first and see if it changes the diff.

Best
Peter

**Comment by soulgalore** [comment]
We've been testing out getting median instead if fastest Speed Index but almost got the same variance. Here we are testing another page:

![windows](https://user-images.githubusercontent.com/540757/36140316-2a7a7784-10a1-11e8-89fb-114da921cdc8.png)

However, I don't see the same big diff between Linux/Windows for the Obama page so maybe it could be that (the diff is 400 points both for Windows & Linux but since Linux is faster the diff in % is much bigger) that the async JavaScript more often kick in and delay the start render. Maybe it is happening more often with the faster Linux. I'll look more into the specific runs to see if I find something.

**Comment by pmeenan** [comment]
I'll see if I can reproduce it in the morning.  If you pull up a couple of runs, can you also compare the RUM first-paint metrics to see if they differ?

If so it is likely a timing thing where sometimes JS hits before it gets a chance to paint and depending on which side of the execution a paint happens it varies.

If it is consistent then it could be something with the alignment of the start time on the video capture (though I'd expect to see the issue across the board then).

FWIW, you can also pass fps=10 as a test option (or any other frame rate) to record video at a different speed.

**Comment by soulgalore** [comment]
Here are a couple of example runs (same series, but larger diff): 
http://wpt.wmftest.org/video/compare.php?tests=180213_YQ_16Y-r:1,180213_YQ_16Y-r:3

And another page: 
http://wpt.wmftest.org/video/compare.php?tests=180213_4D_BK-r%3A1%2C180213_4D_BK-r%3A3&thumbSize=200&ival=100&end=visual#

And the RUM  first paint is almost the same as the video so there's nothing wrong with the video. Maybe it's on our side then, I'll discuss it with my teammates later today.

One thing that is strange is that in both cases the last graph that shows the amount of data differs (but numbers of requests are the same):
<img width="690" alt="screen shot 2018-02-14 at 5 52 18 am" src="https://user-images.githubusercontent.com/540757/36188262-416993b2-114b-11e8-9e91-a9a7fac77753.png">

But when I download the individual HAR files the amount per type is the same between runs, so it seems that there's something else that is picked up?

**Comment by zeman** [comment]
@soulgalore we're seeing size differences with the switch to Linux as well. Pat's planning to take a look shortly. https://github.com/SpeedCurve-Metrics/SpeedCurve/issues/184

**Comment by soulgalore** [comment]
Aha thanks @zeman ! Have you also seen bigger difference between min and max values between runs for Visual Metrics?

**Comment by pmeenan** [comment]
FWIW, I ran through [50 runs on my dev VM](https://www.webpagetest.org/graph_page_data.php?tests=180214_8F_092c24bdc8098bba73fcbbc481bd594c&medianMetric=SpeedIndex#SpeedIndex) for the Sweeden page and there was up to 200ms of variance (30fps capture).

![image](https://user-images.githubusercontent.com/444165/36220178-b19af8a6-1187-11e8-8d59-edf98c9339c9.png)

The first paint and video metrics also aligned and looking at [individual runs](https://www.webpagetest.org/video/compare.php?tests=180214_8F_092c24bdc8098bba73fcbbc481bd594c-r%3A5-c%3A0-l%3A881ms%2C180214_8F_092c24bdc8098bba73fcbbc481bd594c-r%3A23-c%3A0-l%3A999ms&thumbSize=200&ival=16.67&end=visual) it looks like the variance in my tests were from connection set up and SSL negotiation timing differences.  That would show up as variance in TTFB which should be pretty easy to detect.

The examples you had looked different though.  In those cases it looks like timing differences triggered layout and rendering to happen more incrementally in one case than another.

The measurements themselves look accurate (whew) so it's largely going to come down to understanding the source of the timing differences in the actual loading and if they are caused by the VM or just random luck.

**Comment by soulgalore** [comment]
Hi @pmeenan thanks for the spending time on this! Yep 200-300 ms is ok I think we are used to that. The 500 ms is big and the thing is since all pages are different, we usually (when we hit that sweet spot) only got that variance on one of the pages. When I check on Firefox it looks good now (200-300 ms) with the change for ipv6.

Our experience with the AWS c4.large is that it is the best on the market (we have an unpublished blog post in the making where we tried out different providers and AWS introduced the lowest variance of the ones we tested (by far)).

I'll try out different connectivity settings just to see if we got the same behavior then.
